File: src/ai/agent.py
"""
EVOS AI - Agent System
ReAct-style agent that can plan and execute multi-step tasks
"""
import asyncio
import json
from typing import Dict, Any, List, Optional, Callable, Awaitable
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum

from llm_engine import llm_engine
from tools import BrowserTools, ToolResult


class AgentState(str, Enum):
    IDLE = "idle"
    THINKING = "thinking"
    ACTING = "acting"
    WAITING = "waiting"
    COMPLETED = "completed"
    ERROR = "error"


@dataclass
class AgentStep:
    step_number: int
    thought: str
    action: Optional[str] = None
    action_input: Optional[Dict[str, Any]] = None
    observation: Optional[str] = None
    timestamp: datetime = field(default_factory=datetime.now)


@dataclass
class AgentTask:
    task_id: str
    description: str
    context: Dict[str, Any]
    steps: List[AgentStep] = field(default_factory=list)
    state: AgentState = AgentState.IDLE
    result: Optional[Any] = None
    error: Optional[str] = None
    created_at: datetime = field(default_factory=datetime.now)
    completed_at: Optional[datetime] = None


class Agent:
    """
    AI Agent that uses ReAct (Reasoning + Acting) pattern
    to accomplish tasks through tool use.
    """
    
    SYSTEM_PROMPT = """You are EVOS, an intelligent browser agent. You help users accomplish tasks on web pages by using available tools.

You follow the ReAct pattern:
1. THOUGHT: Analyze the current situation and decide what to do next
2. ACTION: Choose and execute a tool
3. OBSERVATION: Analyze the result
4. Repeat until task is complete

Current page context:
URL: {url}
Title: {title}
Page content summary: {content_summary}

Interactive elements on page:
{interactive_elements}

{tools}

IMPORTANT RULES:
- Always think step by step
- Use the most specific tool for each action
- If something fails, try an alternative approach
- When the task is complete, use the "answer_user" tool
- Keep responses concise and focused

Respond in this JSON format:
{{
    "thought": "Your reasoning about what to do",
    "action": "tool_name",
    "action_input": {{"param1": "value1"}}
}}

Or if you're done:
{{
    "thought": "Task completed because...",
    "action": "answer_user",
    "action_input": {{"message": "Here's what I found/did..."}}
}}"""

    def __init__(self, max_steps: int = 10):
        self.max_steps = max_steps
        self.current_task: Optional[AgentTask] = None
        self.tool_executor: Optional[Callable[[str, Dict], Awaitable[ToolResult]]] = None
        
    def set_tool_executor(self, executor: Callable[[str, Dict], Awaitable[ToolResult]]):
        """Set the function that executes tools in the browser"""
        self.tool_executor = executor
    
    async def run(
        self,
        task_description: str,
        context: Dict[str, Any],
        on_step: Optional[Callable[[AgentStep], Awaitable[None]]] = None
    ) -> AgentTask:
        """
        Run the agent to complete a task
        
        Args:
            task_description: What the user wants to accomplish
            context: Current page context (url, title, content, etc.)
            on_step: Callback for each step (for streaming updates)
        
        Returns:
            AgentTask with results
        """
        import uuid
        
        task = AgentTask(
            task_id=str(uuid.uuid4()),
            description=task_description,
            context=context,
            state=AgentState.THINKING
        )
        self.current_task = task
        
        # Get interactive elements from context
        interactive_elements = context.get("interactive_elements", [])
        elements_str = self._format_elements(interactive_elements[:30])  # Limit to 30
        
        # Build system prompt
        system_prompt = self.SYSTEM_PROMPT.format(
            url=context.get("url", "unknown"),
            title=context.get("title", "unknown"),
            content_summary=self._summarize_content(context.get("content", "")),
            interactive_elements=elements_str,
            tools=BrowserTools.format_tools_for_prompt()
        )
        
        # Conversation history for multi-turn
        messages = [
            f"User task: {task_description}\n\nBegin by analyzing the page and planning your approach."
        ]
        
        for step_num in range(1, self.max_steps + 1):
            task.state = AgentState.THINKING
            
            # Build prompt with history
            prompt = "\n\n".join(messages)
            
            # Get LLM response
            response = await llm_engine.generate(
                prompt=prompt,
                system_prompt=system_prompt,
                temperature=0.3,
                max_tokens=1024
            )
            
            # Parse response
            step = self._parse_response(response, step_num)
            task.steps.append(step)
            
            if on_step:
                await on_step(step)
            
            # Check if done
            if step.action == "answer_user":
                task.state = AgentState.COMPLETED
                task.result = step.action_input.get("message", "Task completed")
                task.completed_at = datetime.now()
                break
            
            # Execute action
            if step.action and self.tool_executor:
                task.state = AgentState.ACTING
                
                try:
                    result = await self.tool_executor(step.action, step.action_input or {})
                    step.observation = json.dumps(result.__dict__ if hasattr(result, '__dict__') else result)
                except Exception as e:
                    step.observation = f"Error executing {step.action}: {str(e)}"
                    task.state = AgentState.ERROR
                
                # Add observation to history
                messages.append(f"Step {step_num}:\nThought: {step.thought}\nAction: {step.action}\nAction Input: {json.dumps(step.action_input)}\nObservation: {step.observation}")
            
            # Check for errors
            if step.action is None:
                messages.append(f"Your response wasn't valid JSON. Please respond with proper JSON format.")
        
        if task.state != AgentState.COMPLETED:
            task.state = AgentState.ERROR
            task.error = "Max steps reached without completing task"
        
        return task
    
    def _parse_response(self, response: str, step_num: int) -> AgentStep:
        """Parse LLM response into an AgentStep"""
        step = AgentStep(step_number=step_num, thought="")
        
        try:
            # Find JSON in response
            start = response.find("{")
            end = response.rfind("}") + 1
            
            if start >= 0 and end > start:
                json_str = response[start:end]
                data = json.loads(json_str)
                
                step.thought = data.get("thought", "")
                step.action = data.get("action")
                step.action_input = data.get("action_input", {})
            else:
                step.thought = response
        except json.JSONDecodeError:
            step.thought = response
        
        return step
    
    def _format_elements(self, elements: List[Dict]) -> str:
        """Format interactive elements for the prompt"""
        if not elements:
            return "No interactive elements detected"
        
        lines = []
        for el in elements:
            desc = f"[{el.get('tag', '?')}]"
            if el.get('text'):
                desc += f" '{el['text'][:50]}'"
            if el.get('id'):
                desc += f" id={el['id']}"
            if el.get('name'):
                desc += f" name={el['name']}"
            if el.get('type'):
                desc += f" type={el['type']}"
            if el.get('href'):
                desc += f" -> {el['href'][:50]}"
            lines.append(desc)
        
        return "\n".join(lines)
    
    def _summarize_content(self, content: str, max_length: int = 2000) -> str:
        """Summarize page content for the prompt"""
        if not content:
            return "No content available"
        
        # Clean and truncate
        content = " ".join(content.split())
        if len(content) > max_length:
            content = content[:max_length] + "..."
        
        return content
    
    async def stop(self):
        """Stop the current task"""
        if self.current_task:
            self.current_task.state = AgentState.ERROR
            self.current_task.error = "Task cancelled by user"


class SimpleAgent:
    """
    A simpler agent for quick one-shot tasks
    """
    
    SYSTEM_PROMPT = """You are EVOS, a helpful browser assistant. Analyze the user's request and the current page to provide a helpful response.

Current page:
URL: {url}
Title: {title}
Content: {content}

If the user asks you to do something on the page, respond with a JSON command:
{{"command": "tool_name", "params": {{...}}}}

Available commands: navigate, click, type_text, scroll, extract_text, extract_links

If the user asks a question, just answer it directly based on the page content.

Be concise and helpful."""

    async def quick_response(
        self,
        message: str,
        context: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Get a quick response for simple queries"""
        
        system_prompt = self.SYSTEM_PROMPT.format(
            url=context.get("url", ""),
            title=context.get("title", ""),
            content=context.get("content", "")[:3000]
        )
        
        response = await llm_engine.generate(
            prompt=message,
            system_prompt=system_prompt,
            temperature=0.5,
            max_tokens=512
        )
        
        # Check if response contains a command
        try:
            if "{" in response and "command" in response:
                start = response.find("{")
                end = response.rfind("}") + 1
                data = json.loads(response[start:end])
                return {
                    "type": "command",
                    "command": data.get("command"),
                    "params": data.get("params", {}),
                    "message": response[:start].strip() if start > 0 else ""
                }
        except:
            pass
        
        return {
            "type": "text",
            "message": response
        }


# Singleton instances
agent = Agent()
simple_agent = SimpleAgent()

File: src/ai/agent.js
// EVOS Browser - AI Agent
// ReAct-style agent that uses tools to accomplish tasks

const { llmEngine } = require('./llm-engine');
const { browserTools } = require('./tools');
const { aiMemory } = require('./memory');
const { AGENT_CONFIG } = require('./config');

class AIAgent {
  constructor() {
    this.conversationHistory = [];
    this.currentTask = null;
    this.isProcessing = false;
    this.browserContext = null;
  }

  // Set browser context for tools
  setBrowserContext(context) {
    this.browserContext = context;
    browserTools.setBrowserContext(context);
  }

  // System prompt for the agent
  getSystemPrompt() {
    const toolsDescription = browserTools.getToolDefinitions().map(t => 
      `- ${t.name}: ${t.description}`
    ).join('\n');

    return `You are EVOS AI, an intelligent browser assistant. You help users navigate the web, find information, and automate tasks.

You have access to these tools:
${toolsDescription}

When the user asks you to do something, think step by step:
1. First, understand what the user wants
2. Plan the steps needed to accomplish the task
3. Use the appropriate tools to execute each step
4. Report the results to the user

To use a tool, respond with a JSON block like this:
\`\`\`json
{
  "thought": "I need to search for this information",
  "action": "search_web",
  "params": {
    "query": "your search query"
  }
}
\`\`\`

When you have completed the task or want to respond to the user, use the answer_user tool:
\`\`\`json
{
  "thought": "I have found the information the user requested",
  "action": "answer_user",
  "params": {
    "message": "Here is what I found..."
  }
}
\`\`\`

Important rules:
- Always use tools to interact with web pages - don't make up information
- Be concise and helpful in your responses
- If you encounter an error, try alternative approaches
- Remember context from the conversation
- Always end with answer_user when you're done`;
  }

  // Parse action from LLM response
  parseAction(response) {
    try {
      // Look for JSON in the response
      const jsonMatch = response.match(/```(?:json)?\s*(\{[\s\S]*?\})\s*```/);
      
      if (jsonMatch) {
        const parsed = JSON.parse(jsonMatch[1]);
        return {
          thought: parsed.thought || '',
          action: parsed.action,
          params: parsed.params || {}
        };
      }

      // Try to find raw JSON
      const rawJsonMatch = response.match(/\{[\s\S]*"action"[\s\S]*\}/);
      if (rawJsonMatch) {
        const parsed = JSON.parse(rawJsonMatch[0]);
        return {
          thought: parsed.thought || '',
          action: parsed.action,
          params: parsed.params || {}
        };
      }

      // No action found - treat as direct response
      return {
        thought: 'Responding directly to user',
        action: 'answer_user',
        params: { message: response }
      };
    } catch (error) {
      console.error('[Agent] Failed to parse action:', error);
      return {
        thought: 'Responding directly to user',
        action: 'answer_user',
        params: { message: response }
      };
    }
  }

  // Execute a single step
  async executeStep(userMessage, observations = []) {
    // Build messages for the LLM
    let prompt = '';
    
    // Add conversation context
    const recentHistory = this.conversationHistory.slice(-6);
    for (const msg of recentHistory) {
      prompt += `${msg.role === 'user' ? 'User' : 'Assistant'}: ${msg.content}\n\n`;
    }
    
    // Add current user message if not in history
    if (!recentHistory.some(m => m.content === userMessage && m.role === 'user')) {
      prompt += `User: ${userMessage}\n\n`;
    }
    
    // Add observations from previous actions
    if (observations.length > 0) {
      prompt += 'Previous observations:\n';
      for (const obs of observations) {
        prompt += `- Action: ${obs.action} → Result: ${JSON.stringify(obs.result).substring(0, 500)}\n`;
      }
      prompt += '\n';
    }
    
    prompt += 'Assistant: ';

    // Get memory context
    const memoryContext = await aiMemory.getContext(userMessage);
    
    // Full system prompt
    const fullSystemPrompt = this.getSystemPrompt() + 
      (memoryContext ? `\n\nMemory context:\n${memoryContext}` : '');

    // Generate response
    const response = await llmEngine.chat(
      [{ role: 'user', content: prompt }],
      fullSystemPrompt
    );

    return this.parseAction(response);
  }

  // Run the agent on a task
  async run(userMessage, onProgress = null) {
    if (this.isProcessing) {
      return { error: 'Agent is already processing a task' };
    }

    this.isProcessing = true;
    this.currentTask = userMessage;
    
    // Add user message to history
    this.conversationHistory.push({
      role: 'user',
      content: userMessage,
      timestamp: Date.now()
    });

    const observations = [];
    let step = 0;
    let finalResponse = null;

    try {
      while (step < AGENT_CONFIG.maxSteps) {
        step++;
        
        if (onProgress) {
          onProgress({ type: 'thinking', step, message: 'Thinking...' });
        }

        // Execute step
        const action = await this.executeStep(userMessage, observations);
        
        console.log(`[Agent] Step ${step}:`, action);

        if (onProgress) {
          onProgress({ 
            type: 'action', 
            step, 
            thought: action.thought,
            action: action.action,
            params: action.params
          });
        }

        // Check if this is the final answer
        if (action.action === 'answer_user') {
          finalResponse = action.params.message;
          break;
        }

        // Execute the tool
        try {
          const result = await browserTools.executeTool(action.action, action.params);
          
          observations.push({
            action: action.action,
            params: action.params,
            result: result
          });

          if (onProgress) {
            onProgress({ 
              type: 'observation', 
              step, 
              action: action.action,
              result: result
            });
          }

          // Check if the action result is final
          if (result.final) {
            finalResponse = result.message;
            break;
          }
        } catch (toolError) {
          console.error(`[Agent] Tool error:`, toolError);
          observations.push({
            action: action.action,
            params: action.params,
            result: { error: toolError.message }
          });
        }
      }

      // If we hit max steps without a final answer
      if (!finalResponse) {
        finalResponse = "I've been working on this task but haven't been able to complete it. Could you provide more details or try rephrasing your request?";
      }

      // Add response to history
      this.conversationHistory.push({
        role: 'assistant',
        content: finalResponse,
        timestamp: Date.now()
      });

      // Remember this interaction
      await aiMemory.rememberInteraction(userMessage, finalResponse);

      return {
        success: true,
        response: finalResponse,
        steps: step,
        observations: observations
      };

    } catch (error) {
      console.error('[Agent] Error:', error);
      return {
        success: false,
        error: error.message,
        steps: step,
        observations: observations
      };
    } finally {
      this.isProcessing = false;
      this.currentTask = null;
    }
  }

  // Simple chat without tools
  async chat(message) {
    // Add to history
    this.conversationHistory.push({
      role: 'user',
      content: message,
      timestamp: Date.now()
    });

    // Simple prompt for chat
    const systemPrompt = `You are EVOS AI, a helpful browser assistant built into the EVOS web browser.

When the user's message includes [Current page:...] information, use it to answer questions about what they're viewing.
Be concise, helpful, and accurate. If you see page content, analyze it and respond based on what you can see.
If asked about a page but no content is provided, explain that you need the page content to be extracted first.`;
    
    const response = await llmEngine.chat(
      this.conversationHistory.slice(-10),
      systemPrompt
    );

    // Add response to history
    this.conversationHistory.push({
      role: 'assistant',
      content: response,
      timestamp: Date.now()
    });

    return response;
  }

  // Stream a chat response
  async *chatStream(message) {
    // Add to history
    this.conversationHistory.push({
      role: 'user',
      content: message,
      timestamp: Date.now()
    });

    const systemPrompt = `You are EVOS AI, a helpful browser assistant. Be concise and helpful.`;
    
    // Build prompt
    let prompt = '';
    for (const msg of this.conversationHistory.slice(-10)) {
      prompt += `<|im_start|>${msg.role}\n${msg.content}<|im_end|>\n`;
    }
    prompt += '<|im_start|>assistant\n';

    let fullResponse = '';
    
    for await (const chunk of llmEngine.generateStream(prompt)) {
      fullResponse += chunk;
      yield chunk;
    }

    // Add response to history
    this.conversationHistory.push({
      role: 'assistant',
      content: fullResponse,
      timestamp: Date.now()
    });
  }

  // Clear conversation history
  clearHistory() {
    this.conversationHistory = [];
    llmEngine.resetSession();
  }

  // Get conversation history
  getHistory() {
    return this.conversationHistory;
  }

  // Get agent status
  getStatus() {
    return {
      isProcessing: this.isProcessing,
      currentTask: this.currentTask,
      historyLength: this.conversationHistory.length,
      llmStatus: llmEngine.getInfo()
    };
  }
}

// Singleton instance
const aiAgent = new AIAgent();

module.exports = { AIAgent, aiAgent };
